src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,116 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
+            }
+            else
+            {
+                return (ulong)byteOffset < (ulong)((long)first.Length * Unsafe.SizeOf<T>()) ||
+                       (ulong)byteOffset > (ulong)-((long)second.Length * Unsafe.SizeOf<T>());
+            }
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory and outputs the element offset.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence. If the sequences
+        /// overlap, `elementOffset` is the number of elements by which the second sequence
+        /// is offset from the first sequence (i.e., second minus first). The number is
+        /// rounded away from zero if it is not a whole number, which can happen when
+        /// casting a sequence of a smaller type to a sequence of a larger type. If the
+        /// sequences do not overlap, `elementOffset` is meaningless and arbitrarily set to
+        /// zero.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                if ((uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()))
+                {
+                    elementOffset = ((int)byteOffset + (Unsafe.SizeOf<T>() - 1)) / Unsafe.SizeOf<T>();
```

Comment:
potential for integer overflow?

Reply:
Yes. Dammit.

`byteOffset / SizeOf<T>()` cannot overflow, because a span cannot have more than int.MaxValue elements. But rounding up can result in int.MaxValue+1.

Should I revert to throwing an exception if the spans are misaligned?

Reply:
What about rounding down instead?

I was leaning towards throwing here, but I would defer to Krzysztof.

cc @KrzysztofCwalina

Reply:
Would rounding down, instead of up, work?

Reply:
@ahsonkhan  Rounding in any direction will produce incorrect results if the exact element offset is needed. I some cases, the code doesn't need to know the exact element offset, just (for example) if it is greater than or less than some threshold. In these cases, rounding away from zero produces correct results more often than truncation. See, for example, the included unit tests, which would fail with truncation. So truncation is probably not a good idea. The most "correct" behavior would be an exception.

Reply:
Could have an overload with `System.MidpointRounding` (future)

Reply:
Rounding will never happen in 99.999% of use cases. Users have to specifically craft broken spans with `unsafe` or `NonPortableCast` to trigger this corner case. I don't think the API should force every user of `Overlaps` to consider under which rounding modes their code still works if they are passed broken spans. Silently rounding away from zero would have been nice since it lets *some* code work correctly, but it still leaves other code broken and, as @ahsonkhan noted, can lead to overflow. I don't think rounding has a future. I will push an update that reverts to exceptions.